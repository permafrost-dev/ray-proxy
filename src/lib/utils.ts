import { AxiosResponse } from 'axios';

/**
 * Formats bytes to kilobytes.
 *
 * @param bytes number
 * @param decimals number
 * @returns number
 */
export const formatPayloadSize = (bytes: number, decimals = 3): number => {
    return Number((bytes / 1024.0).toFixed(decimals));
};

/**
 * Sends the response from the Ray app to the proxy client, including headers.
 *  The request flow is as follows:
 *
 *  client --> proxy --> Ray
 *    client <-- proxy <--'
 *
 * @param fastifyReply `any` - reply param from a fastify event handler
 * @param appResponse `any|AxiosResponse` - response from axios request to Ray app
 */
export const relayResponseFromAppToClient = (fastifyReply: any, appResponse: any | AxiosResponse, reflectHeaders = true): void => {
    if (typeof appResponse === 'undefined') {
        const sentReply = fastifyReply.code(404);
        sentReply.header('Access-Control-Allow-Origin', '*');
        sentReply.send('not found');

        return;
    }

    const sentReply = fastifyReply.code(appResponse?.status ?? 404);

    if (reflectHeaders) {
        sentReply.headers(appResponse.headers);
    }

    sentReply.header('connection', 'keep-alive');
    setCorsHeaders(sentReply);

    sentReply.send(appResponse.data);
};

export const setCorsHeaders = (reply: any) => {
    reply.header('Access-Control-Allow-Origin', '*');
    reply.header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS');
    reply.header('Access-Control-Allow-Headers', 'DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range');
    reply.header('Access-Control-Expose-Headers', 'Content-Length,Content-Range');
};

export const sendPreflightCorsResponse = (fastifyReply: any) => {
    const sentReply = fastifyReply.code(204);

    setCorsHeaders(sentReply);

    sentReply.header('Access-Control-Max-Age', 3600 * 24);
    sentReply.header('Content-Type', 'text/plain; charset=utf-8');

    sentReply.send('');
};
