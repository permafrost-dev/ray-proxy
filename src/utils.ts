/**
 * Formats bytes to kilobytes.
 *
 * @param bytes number
 * @param decimals number
 * @returns number
 */
export const formatPayloadSize = (bytes: number, decimals = 3): number => {
    return Number((bytes / 1024.0).toFixed(decimals));
};
