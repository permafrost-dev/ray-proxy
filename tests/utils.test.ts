/* eslint-disable no-unused-vars */
/* eslint-disable no-undef */

import { send } from 'node:process';
import { formatPayloadSize, relayResponseFromAppToClient, sendPreflightCorsResponse, setCorsHeaders } from '../src/lib/utils';

interface SentReplyInterface {
    sentData: any[];
    sentHeaders: any[];

    header(name: string, value: string): void;
    send(data: any): void;
}

let sentReply: SentReplyInterface;

beforeEach(() => {
    sentReply = {
        sentData: [],
        sentHeaders: [],

        header(name: string, value: string) {
            this.sentHeaders.push({ name, value });
        },

        send(data: any) {
            this.sentData.push(data);
        },
    };
});

it('formats bytes to kb', () => {
    expect(formatPayloadSize(1024, 1)).toBe(1.0);
    expect(formatPayloadSize(512, 1)).toBe(0.5);
});

it('relays a response from the app to the client', () => {
    const sentData: any[] = [];
    const sentHeaders: any[] = [];
    let replyCode = 0;

    const sentReply = {
        sentData,
        sentHeaders,

        header(name: string, value: string) {
            this.sentHeaders.push({ name, value });
        },

        headers(obj: any) {
            for (const name in obj) {
                this.header(name, obj[name]);
            }
        },

        send(data: any) {
            this.sentData.push(data);
        },
    };

    const fastifyReply = {
        code(n: any) {
            replyCode = n;
            return sentReply;
        },
    };

    const appResponse = {
        headers: {
            one: 'header one',
            two: 'header two',
        },
        data: {
            abc: 123,
            def: 456,
        },
        status: 200,
    };

    relayResponseFromAppToClient(fastifyReply, appResponse);

    expect(replyCode).toBe(appResponse.status);
    expect(sentData).toMatchSnapshot();
    expect(sentHeaders).toMatchSnapshot();
});

it('sets CORS headers on a reply object', () => {
    setCorsHeaders(sentReply);

    expect(sentReply.sentHeaders).toMatchSnapshot();
});

it('sends preflight response headers', () => {
    let replyCode = 0;

    const fastifyReply = {
        code(n: any) {
            replyCode = n;
            return sentReply;
        },
    };

    sendPreflightCorsResponse(fastifyReply);

    expect(replyCode).toBe(204);
    expect(sentReply.sentHeaders).toMatchSnapshot();
});
