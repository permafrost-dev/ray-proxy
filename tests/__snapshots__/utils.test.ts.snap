// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relays a response from the app to the client 1`] = `
Array [
  Object {
    "abc": 123,
    "def": 456,
  },
]
`;

exports[`relays a response from the app to the client 2`] = `
Array [
  Object {
    "name": "one",
    "value": "header one",
  },
  Object {
    "name": "two",
    "value": "header two",
  },
  Object {
    "name": "connection",
    "value": "keep-alive",
  },
  Object {
    "name": "Access-Control-Allow-Origin",
    "value": "*",
  },
  Object {
    "name": "Access-Control-Allow-Methods",
    "value": "GET, POST, OPTIONS",
  },
  Object {
    "name": "Access-Control-Allow-Headers",
    "value": "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range",
  },
  Object {
    "name": "Access-Control-Expose-Headers",
    "value": "Content-Length,Content-Range",
  },
]
`;

exports[`sends preflight response headers 1`] = `
Array [
  Object {
    "name": "Access-Control-Allow-Origin",
    "value": "*",
  },
  Object {
    "name": "Access-Control-Allow-Methods",
    "value": "GET, POST, OPTIONS",
  },
  Object {
    "name": "Access-Control-Allow-Headers",
    "value": "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range",
  },
  Object {
    "name": "Access-Control-Expose-Headers",
    "value": "Content-Length,Content-Range",
  },
  Object {
    "name": "Access-Control-Max-Age",
    "value": 86400,
  },
  Object {
    "name": "Content-Type",
    "value": "text/plain; charset=utf-8",
  },
]
`;

exports[`sets CORS headers on a reply object 1`] = `
Array [
  Object {
    "name": "Access-Control-Allow-Origin",
    "value": "*",
  },
  Object {
    "name": "Access-Control-Allow-Methods",
    "value": "GET, POST, OPTIONS",
  },
  Object {
    "name": "Access-Control-Allow-Headers",
    "value": "DNT,User-Agent,X-Requested-With,If-Modified-Since,Cache-Control,Content-Type,Range",
  },
  Object {
    "name": "Access-Control-Expose-Headers",
    "value": "Content-Length,Content-Range",
  },
]
`;
